src/validation/validation.cpp

```diff
@@ -1979,10 +1997,28 @@ bool ConnectBlockPrevalidations(const CBlock &block,
 {
     int64_t nTimeStart = GetStopwatchMicros();
 
-    // Check it again in case a previous version let a bad block in
-    if (!CheckBlock(block, state, !fJustCheck, !fJustCheck))
+    // If we're not in IBD then check it again in case a previous version let a bad block in
+    //
+    // During IBD setBlocksAlreadyChecked is used rather than the internal block flag block.fChecked because
+    // blocks are typcially stored 1000 blocks ahead and get writting to disk but fChecked is not stored
+    // to disk, so when the block is finally read again for processing it ends up going through CheckBlock() again
+    // unnecessarily. So by storing the hash of the block during IBD just after the first successful CheckBlock()
+    // we can avoid doing CheckBlock() again later in IBD and which saves us a significant amount of time.
+    //
+    // In addition, if we had just restarted the node and continued IBD, started a Reindex, or found a new block
+    // then setBlockAlreadyChecked would be empty and we'd have to run CheckBlock() at least one time, which is
+    // correct behavior.
     {
-        return false;
+        LOCK(cs_BlocksAlreadyChecked);
+        if (!setBlocksAlreadyChecked.count(block.GetHash()))
+        {
+            if (!CheckBlock(block, state, !fJustCheck, !fJustCheck))
```

Comment:
By putting CheckBlock under the cs_BlocksAlreadyChecked lock it is serialized by this lock.
It would be cleaner if this lock protected exactly what it needs to and no more.

Note that if we are already serialized here by another lock (maybe cs_main), then this will have no substantive effect.  But if we start to parallelize this code and don't fix this now then it'll be something we need to deal with later.

Reply:
good point, i'll update...done.